### PR TITLE
Use upstream MLIR implementation in iree.map_nested_foreach_thread_to_workgroups

### PR DESCRIPTION
**Pull request by grypp**
MLIR's transform dialect has capability to map `scf.foreach_thread` to GPU blocks. [1] adds regarding op `map_nested_foreach_thread_to_gpu_blocks`. [2] moves the op into GPU transform dialect.
    
This PR makes IREE specific transform dialect operation (`iree.map_foreach_thread_to_workgroups`) to make use of the upstream MLIR implementation.

[1] https://reviews.llvm.org/D134190
[2] https://reviews.llvm.org/D134800

### CODE REVIEW

File: compiler/src/iree/compiler/Codegen/Common/TransformExtensions/CommonExtensions.cpp [hunk 1]
```diff
@@ -418,24 +320,72 @@ transform_dialect::ForeachThreadToWorkgroupOp::applyToOne(
...
+  if (failed(maybeTopLevelForeachOp))
     return DiagnosedSilenceableFailure(reportUnknownTransformError(target));
-  }
+  scf::ForeachThreadOp topLevelForeachThreadOp = *maybeTopLevelForeachOp;
 
   SimplePatternRewriter rewriter(topLevelForeachThreadOp);
-  if (failed(rewriteForeachThreadToWorkgroup(topLevelForeachThreadOp, exportOp,
-                                             rewriter)))
+
+  auto generateBlocks = [&](Operation *op, const SmallVector<int64_t> &gridDims,
```
Comment (nicolasvasilache) [inline compiler/src/iree/compiler/Codegen/Common/TransformExtensions/CommonExtensions.cpp:331]: factor this out

**Comment by grypp** [comment]
Should we rename iree op? Currently:
iree -> `foreach_thread_to_workgroup` 
mlir -> `map_nested_foreach_thread_to_gpu_blocks`

For example we could rename it to `map_nested_foreach_thread_to_gpu_workgroups`?

**Comment by nicolasvasilache** [comment]
```
Should we rename iree op? Currently:
iree -> foreach_thread_to_workgroup
mlir -> map_nested_foreach_thread_to_gpu_blocks

For example we could rename it to map_nested_foreach_thread_to_gpu_workgroups?
```

`iree.map_nested_foreach_thread_to_workgroups`  SGTM, note that workgroup is not a GPU-only thing but is reused across all targets by IREE.

Also, please see my comment above and hoist out the lambda into a standalone static func

**Comment by grypp** [comment]
I have updated the PR. It has single commit because previous commits were outdated, so I deleted them.

**Comment by benvanik** [comment]
Closing stale PR.

**Pull request closed without merging**